tensorflow/core/kernels/concat_op.cc

```diff
@@ -369,4 +370,209 @@ REGISTER_KERNEL_BUILDER(Name("ConcatOffset")
+    OP_REQUIRES(c,
+                (0 <= axis && axis < input_dims) ||
+                    (allow_legacy_scalars() && concat_dim == 0),
+                errors::InvalidArgument(
+                    "ConcatOp : Expected concatenating dimensions in the range "
+                    "[",
+                    -input_dims, ", ", input_dims, "), but got ", concat_dim));
+
+    int64_t inputs_flat_dim0 = 1;
+    for (int d = 0; d < axis; ++d) {
+      inputs_flat_dim0 *= input_shape.dim_size(d);
+    }
+    int64_t output_concat_dim = 0;
+    const bool input_is_scalar = IsLegacyScalar(input_shape);
+    int64_t row_size = 0;
+    std::vector<ptrdiff_t> sizes;
+    sizes.reserve(N);
+    std::vector<const SrcT*> input_pointers;
+    input_pointers.reserve(N);
+
+    for (int i = 0; i < N; ++i) {
+      const auto& in = values[i];
+      const bool in_is_scalar = IsLegacyScalar(in.shape());
+      OP_REQUIRES(
+          c, in.dims() == input_dims || (input_is_scalar && in_is_scalar),
+          errors::InvalidArgument(
+              "ConcatOp : Ranks of all input tensors should match: shape[0] = ",
+              input_shape.DebugString(), " vs. shape[", i,
+              "] = ", in.shape().DebugString()));
+      for (int j = 0; j < input_dims; ++j) {
+        if (j == axis) {
+          continue;
+        }
+        OP_REQUIRES(
+            c, in.dim_size(j) == input_shape.dim_size(j),
+            errors::InvalidArgument(
+                "ConcatOp : Dimensions of inputs should match: shape[0] = ",
+                input_shape.DebugString(), " vs. shape[", i,
+                "] = ", in.shape().DebugString()));
+      }
+      if (in.NumElements() > 0) {
+        int64_t inputs_flat_dim1 = in.NumElements() / inputs_flat_dim0;
+        sizes.push_back(inputs_flat_dim1);
+        row_size += sizes.back();
+
+        const SrcT* data_ptr = in.flat<SrcT>().data();
+        input_pointers.push_back(data_ptr);
+      }
+      output_concat_dim += in.dims() > 0 ? in.dim_size(axis) : 1;
+    }
+
+    TensorShape output_shape(input_shape);
+    if (output_shape.dims() == 0) {
+      output_shape.AddDim(output_concat_dim);
+    } else {
+      output_shape.set_dim(axis, output_concat_dim);
+    }
+    Tensor* output = nullptr;
+    OP_REQUIRES_OK(c, c->allocate_output(0, output_shape, &output));
+
+    if (output->NumElements() > 0) {
+      auto output_flat = output->flat<DstT>();
+      DstT* out_ptr = output_flat.data();
+
+      auto worker_threads = c->device()->tensorflow_cpu_worker_threads();
+      int num_threads = std::min(
+          static_cast<int>((output->NumElements() * sizeof(DstT)) / 4096),
+          worker_threads->num_threads);
+      if (num_threads == 1 || num_threads == 0) {
+        for (int64_t i = 0; i < inputs_flat_dim0; ++i) {
+          for (int64_t j = 0; j < N; ++j) {
+            auto size = sizes[j];
+            for (int64_t s = 0; s < size; ++s) {
+              out_ptr[s] = static_cast<DstT>(*input_pointers[j]++);
```

Comment:
@aalbersk  This is inefficient implementation. We could use SIMD code to convert the data type. like dnnl_cvt_float_to_bfloat16() and dnnl_cvt_bfloat16_to_float() API: https://github.com/alibaba/DeepRec/blob/main/third_party/mkl_dnn/oneDNN-v2.3.2-expose-bf16-tp.patch#L53-L55

Reply:
For BF16, oneDNN has given us a quick conversion API.